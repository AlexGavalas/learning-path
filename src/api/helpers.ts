import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
