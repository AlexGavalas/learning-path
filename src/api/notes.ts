import type { AstroComponentFactory } from 'astro/dist/runtime/server';
import { getEntry, render } from 'astro:content';
import flow from 'lodash/fp/flow';
import groupBy from 'lodash/fp/groupBy';
import mapValues from 'lodash/fp/mapValues';

import type { BlogArticleFrontmatter } from '~types/blog';
import type { Note, NoteFrontmatter } from '~types/notes';
import type { SummaryFrontmatter } from '~types/summaries';

import { getNoteMetadata, searchNotes } from './notes-db';

type SearchNotesRpcResponse = {
    title: string;
    line: string;
}[];

type Lines = Record<string, string[]>;

const groupByTitle = flow(
    groupBy<SearchNotesRpcResponse[number]>(({ title }) => title),
    mapValues((notes) => notes.map(({ line }) => line)),
);

export const fetchNotes = async (
    query: string,
): Promise<{
    lines: Lines;
    notes: Note[];
}> => {
    const allNotes = await getNoteMetadata();

    if (query.length > 0) {
        const rows = await searchNotes(query);

        const lines = groupByTitle(rows);

        const noteTitles = new Set(rows.map(({ title }) => title));

        const filteredNotes = allNotes.filter(({ title }) =>
            noteTitles.has(title),
        );

        return {
            lines,
            notes: filteredNotes,
        };
    }

    return {
        lines: {},
        notes: allNotes,
    };
};

export const getNoteData = async ({
    collection,
    slug,
}: {
    collection: 'blog' | 'notes' | 'summaries';
    slug: string;
}): Promise<{
    content: { Content: AstroComponentFactory };
    frontmatter: NoteFrontmatter | SummaryFrontmatter | BlogArticleFrontmatter;
} | null> => {
    const note = await getEntry(collection, slug);

    if (!note) {
        return null;
    }

    return {
        content: await render(note),
        frontmatter: note.data,
    };
};
