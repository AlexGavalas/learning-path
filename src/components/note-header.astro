---
import FormattedDate from '~components/formatted-date.astro';

const { title, created, updated, timeZone } = Astro.props;
---

<h1 class="my-8">{title}</h1>
<div class="mb-4 flex gap-2 text-gray-500">
    <p>
        Created at{' '}
        <FormattedDate dateString={created} timeZone={timeZone} />
    </p>
    <p>/</p>
    <p>
        Updated at{' '}
        <FormattedDate dateString={updated} timeZone={timeZone} />
    </p>
</div>
