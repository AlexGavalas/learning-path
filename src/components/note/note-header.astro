---
import FormattedDate from '~components/formatted-date.astro';
import cn from './note-header.module.css';

interface Props {
    title: string;
    created: string;
    updated: string;
}

const { title, created, updated } = Astro.props;
---

<h1 class={cn.title}>{title}</h1>
<div class={cn.container}>
    <p>
        Created at{' '}
        <FormattedDate dateString={created} />
    </p>
    <span class={cn.separator}>||</span>
    <p>
        Updated at{' '}
        <FormattedDate dateString={updated} />
    </p>
</div>
