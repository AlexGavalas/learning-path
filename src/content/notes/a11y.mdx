---
title: Accessibility
created: '2025-05-26'
updated: '2025-07-24'
published: true
---

Various accessibility resources.

- [A11y Checklist](https://www.a11yproject.com/checklist)
- [A11y Collective](https://www.a11y-collective.com)
- [A11y Myths](https://a11ymyths.com)
- [A11y Not Checklist](https://not-checklist.intopia.digital)
- [A11y Support](https://a11ysupport.io)
- [Access Guide](https://www.accessguide.io)
- [Accessibility Checker](https://www.accessibilitychecker.org)
- [Accessibility for Teams](https://digital.gov/guides/accessibility-for-teams)
- [Accessible Numbers](https://accessiblenumbers.com)
- [APCA Contrast Calculator](https://apcacontrast.com)
- [ARIA Authoring Practices Guide (APG)](https://www.w3.org/WAI/ARIA/apg/patterns)
- [Color A11y](https://stephaniewalter.design/blog/color-accessibility-tools-resources-to-design-inclusive-products)
- [Colour Contrast Checker](https://colourcontrast.cc)
- [Giving a damn about accessibility](https://www.accessibility.uxdesign.cc)
- [pa11y](https://github.com/pa11y/pa11y)
- [Principles of Web Accessibility](https://github.com/Heydon/principles-of-web-accessibility)
- [RTL Styling](https://rtlstyling.com)
- [sa11y](https://github.com/ryersondmp/sa11y)
- [Stark](https://www.getstark.co)
- [ta11y](https://www.ta11y.org)
- [This is WCAG](https://thisiswcag.com)
- [Understanding a11y](https://www.understandingaccessibility.com)
- [WCAG in Plain English](https://aaardvarkaccessibility.com/wcag-plain-english)
- [Web Accessibility: A Reference to Creating Inclusive Websites](https://stackdiary.com/web-accessibility)
- [Web AIM](https://webaim.org)
- [Who Can Use](https://whocanuse.com)

Find more accessibility tools [here](https://github.com/brunopulis/awesome-a11y).
