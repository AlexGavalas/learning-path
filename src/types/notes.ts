import type { CollectionEntry } from 'astro:content';

export type Note = {
    id: string;
    created: string;
    title: string;
    updated: string;
};

export type NoteDBEntry = Note & {
    line: string;
};

export type NotesCollection = CollectionEntry<'notes'>;

export type NoteFrontmatter = NotesCollection['data'];
