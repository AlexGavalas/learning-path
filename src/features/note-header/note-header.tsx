import { FormattedDate } from '~components/formatted-date';

import { type NoteMDX } from '../../../types/notes.types';

type NoteHeaderProps = {
    note: NoteMDX;
};

export const NoteHeader = ({ note }: NoteHeaderProps) => (
    <>
        <h1 className="my-8">{note.title}</h1>
        <div className="mb-4 flex gap-2 text-gray-500">
            <p>
                Created at <FormattedDate dateString={note.created} />
            </p>
            <p> / </p>
            <p>
                Updated at <FormattedDate dateString={note.updated} />
            </p>
        </div>
    </>
);
