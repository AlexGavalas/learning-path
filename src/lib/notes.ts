import matter from 'gray-matter';
import { serialize } from 'next-mdx-remote/serialize';
import rehypeAutolinkHeadings from 'rehype-autolink-headings';
import rehypeExternalLinks from 'rehype-external-links';
import rehypeSlug from 'rehype-slug';

import { supabase } from '~lib/supabase';
import { type NoteMDX } from '~types/notes.types';

export const getAllNoteIds = async () => {
    const { data: fileNames, error } = await supabase.storage
        .from('notes_test_1')
        .list();

    if (error) {
        throw error;
    }

    return fileNames.map((file) => {
        return {
            params: {
                id: file.name.replace(/\.mdx$/, ''),
            },
        };
    });
};

export const getNoteData = async (filename: string): Promise<NoteMDX> => {
    const filePath = `${filename}.mdx`;

    const { data: fileContents, error } = await supabase.storage
        .from('notes_test_1')
        .download(filePath);

    if (error) {
        throw error;
    }

    const matterResult = matter(await fileContents.text());

    const mdxSource = await serialize(matterResult.content, {
        mdxOptions: {
            rehypePlugins: [
                [rehypeExternalLinks, { target: '_blank' }],
                [rehypeSlug],
                [rehypeAutolinkHeadings, { behavior: 'wrap' }],
            ],
        },
    });

    // Can't type gray matter result yet
    return {
        mdxSource,
        ...matterResult.data,
    } as NoteMDX;
};
