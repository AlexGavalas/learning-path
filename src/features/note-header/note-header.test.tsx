import { axe } from 'jest-axe';

import { renderWithUser } from '~test/helpers';

import { type NoteMDX } from '../../../types/notes.types';
import { NoteHeader } from './note-header';

describe('<NoteHeader />', () => {
    const note: NoteMDX = {
        created: '2022-01-01',
        id: 1,
        line: 'line',
        filename: 'filename',
        mdxSource: {
            compiledSource: '',
        },
        title: 'title',
        updated: '2022-01-01',
    };

    it('renders', () => {
        const { container } = renderWithUser(<NoteHeader note={note} />);

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderWithUser(<NoteHeader note={note} />);

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });
});
