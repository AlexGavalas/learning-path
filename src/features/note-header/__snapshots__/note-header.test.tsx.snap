// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoteHeader /> renders 1`] = `
<div>
  <h1
    class="my-8"
  >
    title
  </h1>
  <div
    class="text-gray-500 flex gap-2 mb-4"
  >
    <p>
      Created at 
      <time
        datetime="2022-01-01"
      >
        1 Jan 2022
      </time>
    </p>
    <p>
       / 
    </p>
    <p>
      Updated at 
      <time
        datetime="2022-01-01"
      >
        1 Jan 2022
      </time>
    </p>
  </div>
</div>
`;
